essence-of-live-coding-PortMidi/src/LiveCoding/PortMidi.hs

```diff
@@ -0,0 +1,200 @@
+throwPortMidiC = arrM throwPortMidi
+
+instance MonadTrans PortMidiT where
+  lift = PortMidiT . lift . lift
+
+liftPMError :: Monad m => m (Either PMError a) -> PortMidiT m a
+liftPMError = PortMidiT . ExceptT . fmap (left PMError) . lift
+
+liftHandlingState :: Monad m => Cell (HandlingStateT m) a b -> Cell (PortMidiT m) a b
+liftHandlingState = hoistCell $ PortMidiT . lift
+
+-- | Initialize the MIDI system, run the action, and shut it down again.
+runPortMidiC :: MonadIO m => Cell (PortMidiT m) a b -> CellExcept (HandlingStateT m) a b EOLCPortMidiError
+runPortMidiC cell = try $ proc a -> do
+  _ <- liftCell $ handling portMidiHandle -< ()
+  hoistCell unPortMidiT cell -< a
+
+loopPortMidiC :: MonadIO m => Cell (PortMidiT m) a b -> Cell (HandlingStateT m) a b
+loopPortMidiC cell = foreverC $ runCellExcept $ do
+  e <- runPortMidiC cell
+  once_ $ liftIO $ do
+    putStrLn "Encountered PortMidi exception:"
+    print e
+    threadDelay 1000
+  return e
+
+deriving instance Data PMError
+deriving instance Generic PMError
+instance Finite PMError
+
+-- | A marker witnessing that PortMidi was initialized
+data PortMidiHandle = PortMidiHandle
+
+portMidiHandle :: MonadIO m => Handle m PortMidiHandle
+portMidiHandle = Handle
+  { create = do
+      liftIO initialize
+      return PortMidiHandle
+  , destroy = const $ liftIO $ void terminate
+  }
+
+newtype PortMidiInputStream = PortMidiInputStream { unPortMidiInputStream :: PMStream }
+newtype PortMidiOutputStream = PortMidiOutputStream { unPortMidiOutputStream :: PMStream }
+
+data InputOrOutput = Input | Output
+
+lookupDeviceID :: MonadIO m => String -> InputOrOutput -> m (Either EOLCPortMidiError DeviceID)
+lookupDeviceID nameLookingFor inputOrOutput = do
+  nDevices <- liftIO countDevices
+  -- This is a bit of a race condition, but PortMidi has no better API
+  devices <- forM [0..nDevices-1] $ \deviceID -> do
+    deviceInfo <- liftIO $ getDeviceInfo deviceID
+    return (deviceInfo, deviceID)
+  let allDevicesWithName = filter ((nameLookingFor ==) . name . fst) devices
+      inputDevices = filter (input . fst) allDevicesWithName
+      outputDevices = filter (output . fst) allDevicesWithName
+  return $ case (inputOrOutput, inputDevices, outputDevices) of
+    (_, [], []) -> Left NoSuchDevice
+    (Input, [], _ : _) -> Left NotAnInputDevice
+    (Output, _ : _, []) -> Left NotAnOutputDevice
+    (Input, [(_, deviceID)], _) -> Right deviceID
+    (Output, _, [(_, deviceID)]) -> Right deviceID
+    _ -> Left MultipleDevices
+
+portMidiInputStreamHandle
+  :: MonadIO m
+  => String
+  -> Handle m (Either EOLCPortMidiError PortMidiInputStream)
+portMidiInputStreamHandle name = Handle
+  { create = runExceptT $ do
+      deviceID <- ExceptT $ lookupDeviceID name Input
+      fmap PortMidiInputStream $ withExceptT PMError $ ExceptT $ liftIO $ openInput deviceID
+  -- TODO I don't get the error from closing here.
+  -- Actually I really want ExceptT in the monad
```

Comment:
Follow-up issue

Reply:
I'm not even sure right now what exactly the issue is. I'll let it rest and finish the Handle refactorings first.